### PR TITLE
Features contributions

### PR DESCRIPTION
**Pull request by rait-mediarithmics**
Recipe for bid optimizer model statistics.

### CODE REVIEW

File: app/src/core/bidOptimizer/ViewAllController.js [hunk 1]
```diff
@@ -1,34 +1,73 @@
...
+        '$scope', 'Restangular', 'core/common/auth/Session', '$location', '$uibModal', '$state', '$stateParams', "core/bidOptimizer/PropertyContainer",
+        function ($scope, Restangular, Session, $location, $uibModal, $state, $stateParams, PropertyContainer) {
+            var organisationId = Session.getCurrentWorkspace().organisation_id;
+
+
+            function updateUsedModelId(bidOptimizerId, bidOptimizersUsedModel) {
+
+                Restangular.one('bid_optimizers', bidOptimizerId).all("properties").getList().then(function (properties) {
+
+                    for (var i = 0; i < properties.length; i++) {
```
Comment (dduponchel) [inline app/src/core/bidOptimizer/ViewAllController.js]: I think that 
```js
var latest = properties.filter(function(prop) {
  return prop.technical_name === "latest_model_id" && prop.value.value; 
})[0];
var overriding = properties.filter(function(prop) {
  return prop.technical_name === "overriding_model_id" && prop.value.value; 
})[0];

bidOptimizersUsedModel[bidOptimizerId] = (overriding || latest).value.value;
```
Is shorter and more readable.

File: app/src/core/bidOptimizer/view.all.html [hunk 2]
```diff
@@ -24,12 +24,12 @@
...
       <tr ng-repeat="bidOptimizer in bidOptimizers"
           fetch-plugin-from-version="bidOptimizer.engine_version_id as engine"
-          fetch-plugin-properties="engine.id as engineProperties"
-      >
+          fetch-plugin-properties="engine.id as engineProperties">
         <td>{{bidOptimizer.name}}</td>
         <td>{{engineProperties['name'].value}}</td>
         <td>{{engineProperties['provider'].value}}</td>
         <td class="actions">
+          <a class="mics-btn mics-btn-action" href="#/{{organisationId}}/library/bidOptimizers/{{bidOptimizer.id}}/models/{{bidOptimizersUsedModel[bidOptimizer.id].value}}/report">Details</a>
```
Comment (dduponchel) [inline app/src/core/bidOptimizer/view.all.html]: `bidOptimizersUsedModel` contains the map optimizer id to model id, no need to get the `value` property.

File: app/src/core/bidOptimizer/ViewAllController.js [hunk 3]
```diff
@@ -1,34 +1,73 @@
...
-        });
-      };
-    }
-  ]);
+    'use strict';
+
+    module.controller('core/bidOptimizer/ViewAllController', [
+        '$scope', 'Restangular', 'core/common/auth/Session', '$location', '$uibModal', '$state', '$stateParams', "core/bidOptimizer/PropertyContainer",
+        function ($scope, Restangular, Session, $location, $uibModal, $state, $stateParams, PropertyContainer) {
+            var organisationId = Session.getCurrentWorkspace().organisation_id;
```
Comment (thomas-tosoni) [inline app/src/core/bidOptimizer/ViewAllController.js]: I think that you should store the organisation id directly in the $scope. The variable is useless.

File: app/src/core/bidOptimizer/reports/BidOptimizerReportFeaturesContributionDirective.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                    scope.$watch('report', function (newVal) {
+                        if (newVal) {
+
+                            scope.minSize = 8;
+                            scope.maxSize = 30;
+                            scope.featuresContribution = scope.report.$features_contribution;
+                            var minContribution = 9999;
+                            var maxContribution = -1;
+
+                            for (var idxFeatureContribution in scope.featuresContribution) {
```
Comment (thomas-tosoni) [inline app/src/core/bidOptimizer/reports/BidOptimizerReportFeaturesContributionDirective.js]: What is this for? You are not using categoriesContribution, nor minContribution, nor maxContribution.

File: app/src/core/bidOptimizer/reports/BidOptimizerReportsController.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+define(['./module'], function (module) {
+
+    'use strict';
+
+    module.controller('core/bidOptimizer/reports/BidOptimizerReportsController', [
+        '$scope', 'Restangular', 'core/common/auth/Session', '$stateParams',
+        'core/bidOptimizer/PropertyContainer', '$uibModal', '$resource', 'core/configuration', 'core/common/auth/AuthenticationService',
+        function ($scope, Restangular, Session, $stateParams, PropertyContainer, $uibModal, $resource, configuration, AuthenticationService) {
+
+            var organisationId = Session.getCurrentWorkspace().organisation_id;
```
Comment (thomas-tosoni) [inline app/src/core/bidOptimizer/reports/BidOptimizerReportsController.js]: Same here :v:

File: app/src/core/bidOptimizer/reports/BidOptmizerReportSummaryDataSet.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                            });
+
+                            scope.summary = scope.report.$learning_summary;
+                            scope.learningDataSet = scope.report.$learning_stats.$categories;
+                            scope.validationDataSet = scope.report.$test_stats.$categories;
+
+                            scope.learningDataSetSize = 0;
+                            scope.validationDataSetSize = 0;
+
+                            for (var i = 0; i < scope.learningDataSet.length; i++){
```
Comment (thomas-tosoni) [inline app/src/core/bidOptimizer/reports/BidOptmizerReportSummaryDataSet.js]: With lodash the code becomes sexier :dancer: 
scope.learningDataSetSize = _.reduce(scope.learningDataSet, function(sum, elem) {  return sum + elem.$instance_count }, 0);

File: app/src/core/bidOptimizer/reports/view.report.features-contribution.html [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                     uib-tooltip-template="'feature-contribution'"
+                     tooltip-placement="top"
+                     ng-attr-height="{{maxSize * 2}}"
+                     ng-attr-width="{{maxSize * 2}}">
+                    <circle ng-attr-cx="{{maxSize}}" ng-attr-cy="{{maxSize}}"
+                            ng-attr-r="{{getRadius(featureContribution.$contribution)}}"
+                            ng-attr-fill="{{getColor(featureContribution.$contribution)}}"
+                            class="contribution"></circle>
+                </svg>
+                <script type="text/ng-template" id="feature-contribution">
```
Comment (dduponchel) [inline app/src/core/bidOptimizer/reports/view.report.features-contribution.html]: A template with an id (which *must* be unique) inside a loop is not a good id (same with the other template below).

File: app/src/core/bidOptimizer/reports/view.report.features-contribution.html [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    <table class="ng-scope">
+        <tbody>
+        <tr ng-repeat=" featureContribution in featuresContribution">
+
+            <td class="feature-name">
+                <span>{{featureContribution.$feature_name}}</span>
+
+            </td>
+            <td class="feature-contribution">
+                <svg tooltip-class="detail-tooltip"
```
Comment (dduponchel) [inline app/src/core/bidOptimizer/reports/view.report.features-contribution.html]: You should wrap this `<svg>` inside a `<div>` and let this div handle the tooltip: some browsers has issues with the positioning of the tooltip on the svg. (you will need to add a `display: inline-block;` on this `div[uib-tooltip-template]` to get the same behavior).

File: app/styles/main.scss [hunk 9]
```diff
@@ -1155,6 +1190,209 @@ table.fragment-list-bid-optimizer-engine {
...
+
+    }
+
+    .detail-tooltip .tooltip-inner {
+      @include use-bold-font();
+      padding: 7px;
+      color: $dark-blue;
+      font-size: 13px;
+      font-weight: bold;
+      white-space: nowrap;
```
Comment (dduponchel) [inline app/styles/main.scss]: With long lines you get a strange looking tooltip, I don't think this `nowrap` is a good idea.

File: app/styles/main.scss [hunk 10]
```diff
@@ -1155,6 +1190,209 @@ table.fragment-list-bid-optimizer-engine {
...
+          }
+        }
+        td.actions {
+          @extend .text-right;
+          .mics-btn {
+            @extend .btn-xs;
+            vertical-align: middle;
+          }
+        }
+        /* &:not(:hover) td.actions .mics-btn { */
```
Comment (dduponchel) [inline app/styles/main.scss]: You should `@extend .mics-table;`, not copy/paste it.

**Comment by dduponchel** [comment]
Ok for me.

**Pull request merged**